app/vminsert/netstorage/insert_ctx.go

```diff
@@ -109,16 +118,27 @@ func (ctx *InsertCtx) AddLabel(name, value string) {
 // WriteDataPoint writes (timestamp, value) data point with the given at and labels to ctx buffer.
 func (ctx *InsertCtx) WriteDataPoint(at *auth.Token, labels []prompb.Label, timestamp int64, value float64) error {
 	ctx.MetricNameBuf = storage.MarshalMetricNameRaw(ctx.MetricNameBuf[:0], at.AccountID, at.ProjectID, labels)
-	storageNodeIdx := ctx.GetStorageNodeIdx(at, labels)
-	return ctx.WriteDataPointExt(at, storageNodeIdx, ctx.MetricNameBuf, timestamp, value)
+	storageNodeGroupIDs := ctx.GetStorageNodeGroupIds(at, labels)
+	for _, storageNodeGroupID := range storageNodeGroupIDs {
+		if err := ctx.WriteDataPointExt(at, storageNodeGroupID, ctx.MetricNameBuf, timestamp, value); err != nil {
+			if ctx.AllReplicationGroupsFailed(storageNodeGroupID.Group) {
+				logger.Errorf("All replication groups have failed")
+				return err
+			}
+			logger.Errorf("Replciation group down: %s error: %s", storageNodeGroupID.Group, err)
+		}
+	}
+
+	return nil
 }
 
 // WriteDataPointExt writes the given metricNameRaw with (timestmap, value) to ctx buffer with the given storageNodeIdx.
-func (ctx *InsertCtx) WriteDataPointExt(at *auth.Token, storageNodeIdx int, metricNameRaw []byte, timestamp int64, value float64) error {
-	br := &ctx.bufRowss[storageNodeIdx]
-	sn := storageNodes[storageNodeIdx]
+func (ctx *InsertCtx) WriteDataPointExt(at *auth.Token, sngID StorageNodeGroupID, metricNameRaw []byte, timestamp int64, value float64) error {
+	rg := replicationGroups[sngID.Group]
+	br := &ctx.groupBufRowss[sngID.Group][sngID.Idx]
+	sn := rg.storageNodes[sngID.Idx]
 	bufNew := storage.MarshalMetricRow(br.buf, metricNameRaw, timestamp, value)
-	if len(bufNew) >= maxBufSizePerStorageNode {
+	if len(bufNew) >= rg.maxBufSizePerStorageNode {
```

Comment:
I think it is better from simplicity PoV to have a single global `maxBufSizePerStorageNode` instead of per-replication group value.